Refactored Parallel Code

tanish1729 opened the pull request:
<!--
Welcome to sktime, and thanks for contributing!
Please have a look at our contribution guide:
https://www.sktime.net/en/latest/get_involved/contributing.html
-->

#### Reference Issues/PRs
Works on a part of #5380 
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.
If no issue exists, you can open one here: https://github.com/sktime/sktime/issues
-->


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Updates the code to use `sktime.utils.parallel.parallelize` instead of `joblib Parallel`
There are multiple other instances in the code where such changes have to be made

#### Does your contribution introduce a new dependency? If yes, which one?
No

<!--
Only relevant if you changed pyproject.toml.
We try to minimize dependencies in the core dependency set. There
are also further specific instructions to follow for soft dependencies.
See here for handling dependencies in sktime: https://www.sktime.net/en/latest/developer_guide/dependencies.html
-->

#### What should a reviewer concentrate their feedback on?
- [ ] Correctness of code
<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Did you add any tests for the change?
None, I am still not fully familiar with testing.

<!-- This section is useful if you have added a test in addition to the existing ones. This will ensure that further changes to these files won't introduce the same kind of bug. It is considered good practice to add tests with newly added code to enforce the fact that the code actually works. This will reduce the chance of introducing logical bugs.
-->

#### Any other comments?
<!--
We value all user contributions, no matter how small or complex they are. If you have any questions, feel free to post
in the dev-chat channel on the sktime discord https://discord.com/invite/54ACzaFsn7. If we are slow to review (>3 working days), likewise feel free to ping us on discord. Thank you for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->
Again, not really sure about everything in the checklist but I will be more thorough in future PRs

<!--
Thanks for contributing!
-->

fkiraly commented on sktime/transformations/panel/shapelet_transform.py:
```diff
@@ -38,9 +39,35 @@ class Shapelet:
         The calculated information gain of this shapelet
     data: array-like
         The (z-normalised) data of this shapelet.
+    backend : str, optional
```
I think this is the wrong class.

This should be in `RandomShapeletTransform`, no?

fkiraly reviewed (changes requested):
Thanks!

Two comments:

* you added the `__init__` params in the wrong class, I think this is just an accident
* it would be great if you could add a separate test that checks basic usage for different combinations of backends etc. For that, you can use `_get_parallel_test_fixtures`, and I'd put it in `test_shapelet_transform`.

tanish1729 commented:
Yes that was an error I have fixed that and added the params to the `RandomShapeletTransform` Class.
Can you give me some more info about the 2nd thing related to the test and what I am exactly supposed to do.

Also, I pushed my code to my fork's main branch, I hope you can see the changes reflected too?

fkiraly reviewed (changes requested):
Sure!

Currently, the PR is simply refactored, i.e., the user cannot select different backends.
This is because you do not pass on the `backend` parameters to `parallelize`.

Similarly, in the test you don't pass the backend settings to `RandomShapeletTransform`.

What you should do, is:

* ensure the init parmas get passed into `parallelize`
* ensure in the test the estimator is constructed with those, and make sure that at least locally this runs for you
* handle the deprecation of the old parameters, similar to `ForecastingGridSearchCV` (have a look at older changelogs and versions)

Alternatively, you could, for now, remove the backend parameters from the `__init__` and leave them hard coded internally, and work on deprecation and testing in a second PR.

tanish1729 commented:
The parameters from init are now being passed to parallelize instead of being hardcode. Also, in the test, while creating the `RandomShapeletTransform`, I manually used joblib as the backend.

I am still not sure about how to handle deprecations. Could you elaborate on 
> have a look at older changelogs and versions)

fkiraly commented:
Sure!

We need to tell the user that `n_jobs` will be removed, and the should be using `backend_params` instead.

You could have a look at `Catch22Wrapper` and the todos for version 0.28.0 (for this PR, it should e 0.29.0)

See https://www.sktime.net/en/stable/developer_guide/deprecation.html

fkiraly commented:
let us know if you need any help with the last bit!

Another note, the shapelet module structure hsa been changed, all that has changed is moving the files into a folder further inside. So, if you take your PR and move your files to the same folder (inside `shapelet_transform`), or copy the file content, the conflict should be easily resolvable. Again, let us know if you need help with that.
